File: sdk-ios/Tune/Tune/TunePlaylistManager.m
//
//  TunePlaylistManager.m
//  TuneMarketingConsoleSDK
//
//  Created by Matt Gowie on 8/12/15.
//  Copyright (c) 2015 Tune. All rights reserved.
//

#import "TunePlaylistManager.h"
#import "TuneSkyhookPayload.h"
#import "TuneSkyhookPayloadConstants.h"
#import "TuneApi.h"
#import "TuneManager.h"
#import "TuneFileManager.h"
#import "TuneSkyhookCenter.h"
#import "TuneConfiguration.h"
#import "TuneState.h"
#import "TuneJSONPlayer.h"
#import "TuneCallbackBlock.h"
#import "TuneJSONUtils.h"

@interface TunePlaylistManager ()

@property (strong, nonatomic) TunePlaylist *currentPlaylist;
@property (assign, nonatomic) BOOL isUpdating;

@end

@implementation TunePlaylistManager

NSTimer *playlistScheduler;
static BOOL startedScheduledDispatch = NO;
static BOOL receivedFirstPlaylistDownload = NO;

static BOOL gotVeryFirstPlaylist = NO;
TuneCallbackBlock *firstPlaylistDownloadedBlock;
NSObject *firstPlaylistDownloadedBlockLock;
NSOperationQueue *playlistCallbackQueue;

#pragma mark - Initialization / Deallocation

-(id)initWithTuneManager:(TuneManager *)tuneManager {
    self = [super initWithTuneManager:tuneManager];
    
    if (self) {
        playlistCallbackQueue = [NSOperationQueue new];
        firstPlaylistDownloadedBlockLock = [[NSObject alloc] init];
        
        startedScheduledDispatch = NO;
        receivedFirstPlaylistDownload = NO;
        gotVeryFirstPlaylist = NO;
    }
    
    return self;
}

- (void)bringUp {
    [self registerSkyhooks];
    startedScheduledDispatch = NO;
}

- (void)bringDown {
    [self unregisterSkyhooks];
    [playlistScheduler invalidate];
    playlistScheduler = nil;
}

#pragma mark - Skyhook registration

- (void)registerSkyhooks {
    [self unregisterSkyhooks];
    [[TuneSkyhookCenter defaultCenter] addObserver:self selector:@selector(didEnterForegroundSkyhook:) name:TuneSessionManagerSessionDidStart object:nil priority:TuneSkyhookPriorityFirst];
    [[TuneSkyhookCenter defaultCenter] addObserver:self selector:@selector(didEnterBackgroundSkyhook:) name:TuneSessionManagerSessionDidEnd object:nil priority:TuneSkyhookPriorityFirst];
    [[TuneSkyhookCenter defaultCenter] addObserver:self selector:@selector(playlistProcessed:) name:TunePlaylistUpdatePlaylist object:nil priority:TuneSkyhookPriorityFirst];
    [[TuneSkyhookCenter defaultCenter] addObserver:self selector:@selector(handleOnFirstPlaylistDownloaded:) name:TunePlaylistManagerFirstPlaylistDownloaded object:nil priority:TuneSkyhookPriorityFirst];
}

#pragma mark - Skyhook Calls

- (void)didEnterForegroundSkyhook:(TuneSkyhookPayload*)payload {
    if ([TuneState isTMADisabled]) { return; }
    
    // If there's a callback holder that was canceled on app background,
    // restart it with timeout
    if (firstPlaylistDownloadedBlock != nil && [firstPlaylistDownloadedBlock isCanceled]) {
        [firstPlaylistDownloadedBlock startTimer:[firstPlaylistDownloadedBlock getDelay]];
    }
    
    [self fetchAndUpdatePlaylist];
    
    if (!startedScheduledDispatch) {
        NSTimeInterval interval = [self.tuneManager.configuration.playlistRequestPeriod integerValue];
        if (self.tuneManager.configuration.pollForPlaylist) {
            playlistScheduler = [NSTimer scheduledTimerWithTimeInterval:interval target:self selector:@selector(fetchAndUpdatePlaylist) userInfo:nil repeats:YES];
            startedScheduledDispatch = YES;
        }
    }
}

- (void)didEnterBackgroundSkyhook:(TuneSkyhookPayload*)payload {
    if ([TuneState isTMADisabled]) { return; }
    
    // Reset received first playlist flag so we fire the first Playlist Skyhook on next download
    receivedFirstPlaylistDownload = NO;
    
    if (startedScheduledDispatch) {
        [playlistScheduler invalidate];
        playlistScheduler = nil;
        startedScheduledDispatch = NO;
    }
    
    // If there's a currently pending callback, stop it upon background
    if (firstPlaylistDownloadedBlock != nil) {
        [firstPlaylistDownloadedBlock stopTimer];
    }
}

#pragma mark - Fetch Playlist

- (void)fetchAndUpdatePlaylist {
    InfoLog(@"Fetch and Update Playlist!");
    
    // do not allow concurrent updates
    if (self.isUpdating) { return; }
    
    self.isUpdating = YES;
    
    TuneHttpRequest *request = [TuneApi getPlaylistRequest];
    
    // If there was an error building the request, then don't do anything
    if (request == nil) {
        self.isUpdating = NO;
        return;
    }
    
    request.timeoutInterval = 15.0;
    __block TunePlaylistManager *_self = self;
    [request performAsynchronousRequestWithCompletionBlock:^(TuneHttpResponse *response) {
        @try {
            if ([response error]) {
                WarnLog(@"Error downloading playlist from %@, Error: %@", request.URL, [[response error] localizedDescription]);
            } else {
                InfoLog(@"Successfully downloaded the playlist.");
            }
            
            NSDictionary *playlistDictionary = nil;
            if (_self.tuneManager.configuration.usePlaylistPlayer) {
                playlistDictionary = [_self.tuneManager.playlistPlayer getNext];
            } else if (response.wasSuccessful) {
                playlistDictionary = response.responseDictionary;
            }
            
            TunePlaylist *newPlaylist = nil;
            if (playlistDictionary == nil) {
                WarnLog(@"Playlist response did not have any JSON");
                // If we failed to get the playlist (or if it is empty) then trigger the 'onFirstPlaylistDownloaded' callback immediately
                // So that we don't hit the timeout.
                [self handleOnFirstPlaylistDownloaded:nil];
            } else if (playlistDictionary.count == 0) {
                /*
                 *  IMPORTANT:
                 *      An empty playlist is a signal from the server to not process anything
                 */
                WarnLog(@"Received empty playlist from the server -- not updating");
                [self handleOnFirstPlaylistDownloaded:nil];
            } else {
                newPlaylist = [TunePlaylist playlistWithDictionary:playlistDictionary];
                
                if ([TuneManager currentManager].configuration.echoPlaylists) {
                    NSLog(@"Got playlist:\n%@", [TuneJSONUtils createPrettyJSONFromDictionary:playlistDictionary withSecretTMADepth:nil]);
                }
            }
            
            if (newPlaylist) {
                // Save the new playlist
                [[TuneSkyhookCenter defaultCenter] postSkyhook:TunePlaylistUpdatePlaylist object:newPlaylist userInfo:nil];
            } else {
                // Even if the playlist manager failed to download and no new playlist was returned, we still post the Finished Download Skyhook.
                [[TuneSkyhookCenter defaultCenter] postSkyhook:TunePlaylistManagerFinishedPlaylistDownload object:self];
            }
        } @catch (NSException *exception) {
            ErrorLog(@"Error processing the playlist: %@", exception);
        } @finally {
            _self.isUpdating = NO;
        }
    }];
}

- (void)playlistProcessed:(TuneSkyhookPayload *)payload {
    TunePlaylist *processedPlaylist = payload.object;
    if (processedPlaylist) {
        [self setCurrentPlaylist:processedPlaylist];
    }
    
    [[TuneSkyhookCenter defaultCenter] postSkyhook:TunePlaylistManagerFinishedPlaylistDownload object:self];
}

#pragma mark - Set Current Playlist

- (void)setCurrentPlaylist:(TunePlaylist *)newPlaylist {
    // If TMA is disabled, the new playlist is always blank.
    if ([TuneState isTMADisabled]) {
        newPlaylist = [TunePlaylist playlistWithDictionary:@{}];
    }

    if (_currentPlaylist == nil || ![_currentPlaylist isEqual:newPlaylist]) {
        _currentPlaylist = newPlaylist;
        
        // Only save this playlist if it is not from disk and it is not from connected mode
        // This could be our first install playlist download or a new playlist that is different than the one on disk.
        if (!newPlaylist.fromDisk && !newPlaylist.fromConnectedMode) {
            [TuneFileManager savePlaylistToDisk:newPlaylist];
        }
        
        NSMutableDictionary *userInfo = @{}.mutableCopy;
        if (newPlaylist) {
            userInfo[TunePayloadNewPlaylist] = newPlaylist;
            userInfo[TunePayloadPlaylistLoadedFromDisk] = @(newPlaylist.fromDisk);
        }
        
        // Allow internal state to be updated after playlist download, e.g. power hooks
        [[TuneSkyhookCenter defaultCenter] postSkyhook:TunePlaylistManagerCurrentPlaylistChanged object:self userInfo:userInfo];
    }
    
    // Once the internal state is updated, if this is the first playlist we've downloaded this session then send off the FirstPlaylistDownloaded hook
    if (!newPlaylist.fromDisk && !receivedFirstPlaylistDownload) {
        receivedFirstPlaylistDownload = YES;
        [[TuneSkyhookCenter defaultCenter] postSkyhook:TunePlaylistManagerFirstPlaylistDownloaded object:self userInfo:@{ TunePayloadFirstPlaylistDownloaded:newPlaylist }];
    }
}

#pragma mark - Load / Save Playlist to Disk

- (BOOL)loadPlaylistFromDisk {
    NSDictionary *playlistFromDisk;
    
    if (self.tuneManager.configuration.usePlaylistPlayer) {
        playlistFromDisk = [self.tuneManager.playlistPlayer getNext];
    } else {
        playlistFromDisk = [TuneFileManager loadPlaylistFromDisk];
    }
    
    if (playlistFromDisk) {
        TunePlaylist *playlist = [TunePlaylist playlistWithDictionary:playlistFromDisk];
        playlist.fromDisk = YES;
        [self setCurrentPlaylist:playlist];
        
        return YES;
    } else {
        return NO;
    }
}

#pragma mark - On First Playlist Downloaded Callbacks

- (void)onFirstPlaylistDownloaded:(void (^)(void))block withTimeout:(NSTimeInterval)timeout {
    @try {
        TuneCallbackBlock *blockCallback = [[TuneCallbackBlock alloc] initWithCallbackBlock:block fireOnce:YES];
        
        BOOL executeBlock = NO;
        
        @synchronized(firstPlaylistDownloadedBlockLock) {
            // If there's a currently pending callback, stop it
            if (firstPlaylistDownloadedBlock != nil) {
                [firstPlaylistDownloadedBlock stopTimer];
            }
            
            if (gotVeryFirstPlaylist || [TuneState isTMADisabled]) {
                executeBlock = YES;
            } else {
                if (timeout > 0) {
                    [blockCallback startTimer:timeout];
                }
                firstPlaylistDownloadedBlock = blockCallback;
            }
        }
        
        if (executeBlock) {
            [playlistCallbackQueue addOperationWithBlock:^{
                [blockCallback executeBlock];
            }];
        }
    } @catch (NSException *exception) {
        ErrorLog(@"Error in onFirstPlaylistDownloaded: %@", exception);
    }
}

- (void)handleOnFirstPlaylistDownloaded:(TuneSkyhookPayload *)payload {
    @synchronized(firstPlaylistDownloadedBlockLock) {
        if (!gotVeryFirstPlaylist) {
            gotVeryFirstPlaylist = YES;
            
            if (firstPlaylistDownloadedBlock != nil) {
                [playlistCallbackQueue addOperationWithBlock:^{
                    [firstPlaylistDownloadedBlock executeBlock];
                }];
            }
        }
    }
}

#pragma mark - User in Segment API

- (BOOL)isUserInSegmentId:(NSString *)segmentId {
    // If nil or empty string, return false
    if ([segmentId length] == 0) {
        return NO;
    }
    
    // If the segments dictionary from the playlist has a value for the segment id, then user is in this segment
    return [_currentPlaylist.segments objectForKey:segmentId] != nil;
}

- (BOOL)isUserInAnySegmentIds:(NSArray<NSString *> *)segmentIds {
    // If nil input, return false
    if (!segmentIds) {
        return NO;
    }
    
    NSSet* inputSegmentIds = [NSSet setWithArray:segmentIds];
    NSSet* playlistSegmentIds = [NSSet setWithArray:[_currentPlaylist.segments allKeys]];
    return [inputSegmentIds intersectsSet:playlistSegmentIds]; // Return whether any segment ids are found in both sets
}

- (void)forceSetUserInSegment:(BOOL)isInSegment forSegmentId:(NSString *)segmentId {
    // Get NSMutableDictionary of playlist's segments so we can add our force set value
    NSMutableDictionary *modifiedSegments = [NSMutableDictionary dictionaryWithDictionary:_currentPlaylist.segments];
    
    // If user should be in segment, add them to the playlist's segments dictionary
    if (isInSegment) {
        [modifiedSegments setValue:@"Set by forceSetUserInSegmentId" forKey:segmentId];
    } else {
        // Else remove any existing segment key-values
        [modifiedSegments removeObjectForKey:segmentId];
    }
    
    // Convert the NSMutableDictionary back to NSDictionary and set it as the playlist's segments
    _currentPlaylist.segments = [NSDictionary dictionaryWithDictionary:modifiedSegments];
}

@end

File: sdk-android/TuneMarketingConsoleSDK/src/main/java/com/tune/ma/playlist/TunePlaylistManager.java
package com.tune.ma.playlist;

import android.text.TextUtils;

import com.tune.TuneDebugLog;
import com.tune.ma.TuneManager;
import com.tune.ma.configuration.TuneConfigurationManager;
import com.tune.ma.eventbus.TuneEventBus;
import com.tune.ma.eventbus.event.TuneAppBackgrounded;
import com.tune.ma.eventbus.event.TuneAppForegrounded;
import com.tune.ma.eventbus.event.TunePlaylistManagerCurrentPlaylistChanged;
import com.tune.ma.eventbus.event.TunePlaylistManagerFirstPlaylistDownloaded;
import com.tune.ma.model.TuneCallback;
import com.tune.ma.model.TuneCallbackHolder;
import com.tune.ma.playlist.model.TunePlaylist;
import com.tune.ma.utils.TuneJsonUtils;

import org.greenrobot.eventbus.Subscribe;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Created by gowie on 1/27/16.
 * @deprecated IAM functionality. This method will be removed in Tune Android SDK v6.0.0
 */
@Deprecated
public class TunePlaylistManager {

    private TuneCallbackHolder onFirstPlaylistDownloadCallbackHolder;
    private boolean firstPlaylistCallbackExecuted;

    private TunePlaylist currentPlaylist;
    private boolean started;
    private boolean isUpdating;
    private boolean receivedFirstPlaylistDownload;

    private ScheduledThreadPoolExecutor scheduler;
    private ExecutorService executorService;

    private static final String TAG = "PlaylistManager";
    
    public TunePlaylistManager() {
        this.started = false;
        this.isUpdating = false;
        this.receivedFirstPlaylistDownload = false;
        this.executorService = Executors.newSingleThreadExecutor();
        this.loadPlaylistFromDisk();
    }

    //  EventBus Events
    @Subscribe(priority = TuneEventBus.PRIORITY_FIFTH)
    public synchronized void onEvent(TuneAppForegrounded event) {
        if (TuneManager.getInstance() == null || TuneManager.getInstance().getConfigurationManager().isTMADisabled()) {
            return;
        }

        // If there's a callback holder that was canceled on app background, resume it
        if (onFirstPlaylistDownloadCallbackHolder != null && onFirstPlaylistDownloadCallbackHolder.isCanceled()) {
            // If playlist is downloaded, execute callback immediately
            if (receivedFirstPlaylistDownload) {
                onFirstPlaylistDownloadCallbackHolder.executeBlock();
            } else {
                long timeout = onFirstPlaylistDownloadCallbackHolder.getTimeout();
                if (timeout > 0) {
                    // Restart the canceled callback with timeout
                    onFirstPlaylistDownloadCallbackHolder.setTimeout(timeout);
                }
            }
        }

        if (TuneManager.getInstance().getConfigurationManager().getPollForPlaylist() && !this.started) {
            this.startPlaylistRetriever();
            started = true;
        } else {
            executorService.execute(new PlaylistRetrievalTask());
        }
    }

    @Subscribe(priority = TuneEventBus.PRIORITY_FIFTH)
    public synchronized void onEvent(TuneAppBackgrounded event) {
        if (started) {
            this.stopPlaylistRetriever();
            started = false;
        }

        // If there's a currently pending callback holder, stop it upon background
        if (onFirstPlaylistDownloadCallbackHolder != null) {
            onFirstPlaylistDownloadCallbackHolder.stopTimer();
        }
    }

    // Start/Stop Retriever Task
    /////////////////////////////
    private void startPlaylistRetriever() {
        TuneDebugLog.i(TAG, "Starting PlaylistRetriever Schedule.");
        scheduler = new ScheduledThreadPoolExecutor(1);
        scheduler.scheduleAtFixedRate(new PlaylistRetrievalTask(), 0, TuneManager.getInstance().getConfigurationManager().getPlaylistRequestPeriod(), TimeUnit.SECONDS);
    }

    private void stopPlaylistRetriever() {
        if (scheduler != null) {
            TuneDebugLog.i(TAG, "Stopping PlaylistRetriever Schedule.");
            scheduler.shutdownNow();
            scheduler = null;
        }
    }

    private boolean loadPlaylistFromDisk() {
        JSONObject playlistFromDisk = null;
        if (TuneManager.getInstance().getConfigurationManager().usePlaylistPlayer()) {
            playlistFromDisk = TuneManager.getInstance().getPlaylistPlayer().getNext();
        } else {
            playlistFromDisk = TuneManager.getInstance().getFileManager().readPlaylist();
        }

        if (playlistFromDisk != null) {
            TunePlaylist playlist = new TunePlaylist(playlistFromDisk);
            playlist.setFromDisk(true);
            setCurrentPlaylist(playlist);
            return true;
        } else {
            return false;
        }
    }

    public synchronized boolean isUserInSegmentId(String segmentId) {
        // If segment id is null or empty, return false
        if (TextUtils.isEmpty(segmentId)) {
            return false;
        }

        JSONObject segments = currentPlaylist.getSegments();
        // Segments key wasn't found in playlist, return false
        if (segments == null) {
            return false;
        }

        JSONArray segmentIds = segments.names();

        // Segments was empty, return false
        if (segmentIds == null) {
            return false;
        }

        // Check if playlist segment ids contains the one in question
        for (int i = 0; i < segmentIds.length(); i++) {
            if (segmentIds.optString(i).equals(segmentId)) {
                return true;
            }
        }
        return false;
    }

    public synchronized boolean isUserInAnySegmentIds(List<String> segmentIdsToCheck) {
        // Input is null, return false
        if (segmentIdsToCheck == null) {
            return false;
        }

        JSONObject segments = currentPlaylist.getSegments();
        // Segments key wasn't found in playlist, return false
        if (segments == null) {
            return false;
        }

        JSONArray segmentIds = segments.names();

        // Segments was empty, return false
        if (segmentIds == null) {
            return false;
        }

        // Convert JSONArray to List
        List<String> segmentIdsList = new ArrayList<String>();
        for (int i = 0; i < segmentIds.length(); i++) {
            segmentIdsList.add(segmentIds.optString(i));
        }

        // Return whether the two lists share any elements in common
        return (!Collections.disjoint(segmentIdsList, segmentIdsToCheck));
    }

    public synchronized void forceSetUserInSegmentId(String segmentId, boolean isInSegment) {
        // If segment key wasn't found in playlist, playlist segments is null
        // Initialize it so that we can add dummy values
        if (currentPlaylist.getSegments() == null) {
            currentPlaylist.setSegments(new JSONObject());
        }

        try {
            // If true, add it with a dummy value to segments
            if (isInSegment) {
                currentPlaylist.getSegments().put(segmentId, "Set by forceSetUserInSegmentId");
            } else {
                // Otherwise, try to remove it from segments
                currentPlaylist.getSegments().remove(segmentId);
            }
        } catch (JSONException e) {
        }
    }

    public synchronized void getConnectedPlaylist() {
        if (TuneManager.getInstance().getConfigurationManager().getPollForPlaylist() && this.started) {
            stopPlaylistRetriever();
            this.started = false;
        }
        executorService.execute(new PlaylistRetrievalTask());
    }

    public TunePlaylist getCurrentPlaylist() {
        return currentPlaylist;
    }

    protected synchronized void setCurrentPlaylist(TunePlaylist newPlaylist) {
        // If TMA is disabled, the new playlist is always blank.
        if (TuneManager.getInstance() == null || TuneManager.getInstance().getConfigurationManager().isTMADisabled()) {
            newPlaylist = new TunePlaylist();
        }

        if (currentPlaylist == null || !currentPlaylist.equals(newPlaylist)) {
            this.currentPlaylist = newPlaylist;

            // Current playlist changed and not from disk, save new playlist to disk
            if (TuneManager.getInstance() != null && !newPlaylist.isFromDisk() && !newPlaylist.isFromConnectedMode()) {
                TuneDebugLog.i("Saving New Playlist to Disk");
                TuneManager.getInstance().getFileManager().writePlaylist(currentPlaylist.toJson());
            }

            if (TuneManager.getInstance() != null) {
                // Update power hooks directly, EventBus may not be available yet
                TuneManager.getInstance().getPowerHookManager().updatePowerHooksFromPlaylist(currentPlaylist);
            }

            TuneEventBus.post(new TunePlaylistManagerCurrentPlaylistChanged(newPlaylist));
        }

        if (!newPlaylist.isFromDisk()) {
            checkTriggerOnFirstPlaylistDownloadedCallback();
        }
    }

    private class PlaylistRetrievalTask implements Runnable {

        @Override
        public void run() {
            // If we are off then don't bother requesting anymore playlists
            if (TuneManager.getInstance() == null || TuneManager.getInstance().getConfigurationManager().isTMADisabled()) {
                return;
            }

            TuneDebugLog.i("Retrieving Playlist from Server");

            if (isUpdating) {
                return;
            }

            isUpdating = true;

            try {
                TuneConfigurationManager configurationManager = TuneManager.getInstance().getConfigurationManager();

                boolean fromConnectedMode = false;
                JSONObject response = null;
                if (configurationManager.usePlaylistPlayer()) {
                    response = TuneManager.getInstance().getPlaylistPlayer().getNext();
                } else if (TuneManager.getInstance().getConnectedModeManager().isInConnectedMode()){
                    response = TuneManager.getInstance().getApi().getConnectedPlaylist();
                    fromConnectedMode = true;
                } else {
                    response = TuneManager.getInstance().getApi().getPlaylist();
                }

                TunePlaylist newPlaylist = null;
                if (response == null) {
                    TuneDebugLog.w("Playlist response did not have any JSON");
                    // If we failed to get the playlist (or if it is empty) then trigger the 'onFirstPlaylistDownloaded' callback immediately
                    // So that we don't hit the timeout.
                    checkTriggerOnFirstPlaylistDownloadedCallback();
                } else if (response.length() == 0) {
                    /*
                     *  IMPORTANT:
                     *     An empty playlist is a signal from the server to not process anything
                     */

                    TuneDebugLog.w("Received empty playlist from the server -- not updating");
                    checkTriggerOnFirstPlaylistDownloadedCallback();
                } else {
                    if (configurationManager.echoPlaylists()) {
                        TuneDebugLog.alwaysLog("Got Playlist:\n" + TuneJsonUtils.getPrettyJson(response));
                    }
                    newPlaylist = new TunePlaylist(response);
                    newPlaylist.setFromConnectedMode(fromConnectedMode);
                }

                if (newPlaylist != null) {
                    setCurrentPlaylist(newPlaylist);
                }
            } catch (Exception e) {
                e.printStackTrace();
                TuneDebugLog.e(TAG, "Failed to download new playlist.");
            } finally {
                isUpdating = false;
            }
        }
    }

    // On First Playlist Downloaded Callback
    //////////////////////////////////////////

    public synchronized void onFirstPlaylistDownloaded(final TuneCallback callback, long timeout) {
        if (callback == null) {
            TuneDebugLog.IAMConfigError("You passed a null TuneCallback for the onFirstPlaylistDownloaded callback.");
            return;
        }

        // If there's a currently pending callback holder, stop it
        if (onFirstPlaylistDownloadCallbackHolder != null) {
            onFirstPlaylistDownloadCallbackHolder.stopTimer();
        }

        // Initialize callback executed to false, since this method is registering it
        setFirstPlaylistCallbackExecuted(false);
        if (TuneManager.getInstance() == null || TuneManager.getInstance().getConfigurationManager().isTMADisabled()) {
            TuneDebugLog.i("TMA is Disabled, executing firstPlaylistDownload callback");
            setFirstPlaylistCallbackExecuted(true);
            executorService.execute(new Runnable() {
                @Override
                public void run() {
                    callback.execute();
                }
            });
        } else if (receivedFirstPlaylistDownload) {
            TuneDebugLog.i("Playlist already downloaded upon callback registration, executing firstPlaylistDownload callback");
            setFirstPlaylistCallbackExecuted(true);
            executorService.execute(new Runnable() {
                @Override
                public void run() {
                    callback.execute();
                }
            });
        } else {
            onFirstPlaylistDownloadCallbackHolder = new TuneCallbackHolder(callback);

            if (timeout > 0) {
                TuneDebugLog.i("Playlist not downloaded, executing firstPlaylistDownload callback after timeout " + timeout);
                onFirstPlaylistDownloadCallbackHolder.setTimeout(timeout);
            }
        }
    }

    public boolean hasFirstPlaylistCallbackExecuted() {
        return firstPlaylistCallbackExecuted;
    }

    public void setFirstPlaylistCallbackExecuted(boolean firstPlaylistCallbackExecuted) {
        this.firstPlaylistCallbackExecuted = firstPlaylistCallbackExecuted;
    }

    private synchronized void checkTriggerOnFirstPlaylistDownloadedCallback() {
        if (!receivedFirstPlaylistDownload) {
            receivedFirstPlaylistDownload = true;

            // If we have a pending callback that hasn't been executed, execute it
            if (!firstPlaylistCallbackExecuted && onFirstPlaylistDownloadCallbackHolder != null && !onFirstPlaylistDownloadCallbackHolder.isCanceled()) {
                TuneDebugLog.i("Playlist downloaded, executing firstPlaylistDownload callback");
                try {
                    executorService.execute(new Runnable() {
                        @Override
                        public void run() {
                            onFirstPlaylistDownloadCallbackHolder.executeBlock();
                        }
                    });
                } catch (Exception e) {
                    TuneDebugLog.e("Exception in executing firstPlaylistDownload callback.", e);
                }
            }

            // Send EventBus event for first playlist downloaded
            TuneEventBus.post(new TunePlaylistManagerFirstPlaylistDownloaded());
        }
    }
}
